Jetpack Offer Reset: Show product options in product description (daily or real-time).

elliottprogrammer opened the pull request:
#### Changes proposed in this Pull Request

This PR adds the text, "Available options: Real-Time and Daily" to the product description, IF:
- The product offers a real-time and daily option.
- and if the product is not already owned by the user.

#### Testing instructions

- Ensure your sandbox site does't have any Jetpack plans/products purchased.
- Enable the _Offer Reset Flow_ in _A/B tests_.
- Visit the Plans section.

Look at the description section in each of the Plan/Individual Product cards.  Every product/plan that offers a "real-time" and "daily" option will have, "Available options: Real-Time or Daily." appended to the description.  In this case, "Jetpack Backup" and "Jetpack Security" with have the appended text because those are the only plans that offer real-time and daily options at this point.

Screenshots:
![image](https://user-images.githubusercontent.com/11078128/90275283-df95ba80-de16-11ea-905f-116531abccb8.png)

![image](https://user-images.githubusercontent.com/11078128/90275694-8da16480-de17-11ea-917d-c44ebc216139.png)
> Please note:  The "Jetpack Backup" product only has some Lorem Ipsum for the product description. I'm assuming this will be handled in another PR.

Next, purchase Jetpack Backup, reload the Plans page and You should no longer see the appended text ("Available options: Real-Time or Daily.") in the product description of the Jetpack Backup product card. 

Screenshot:
![image](https://user-images.githubusercontent.com/11078128/90276476-d4438e80-de18-11ea-8389-4ce45f87e35b.png)

monsieur-z commented:
The `JetpackProductCard` `description` prop accepts a React node. I believe we could add the available options in the description to save a prop.

rcanepa commented:
@elliottprogrammer, since you started this PR things have changed quite a bit in a couple of these files. Can you update your PR and ping us when it's ready for a new review? 

Advice: always include Jetpack Voyager as reviewer so everyone can be aware of it.

rcanepa commented on client/my-sites/plans-v2/constants.ts:
```diff
@@ -102,7 +102,7 @@ export const OPTION_PRODUCT_BACKUP: SelectorProduct = {
 	displayName: translate( 'Jetpack Backup' ),
 	tagline: '',
-	description: '',
+	description: 'Lorem ipsum',
```
Since we are close to having a version that can be opened for internal testing I think it isn't a good idea to add placeholder text (we might forget to delete/change it). In https://wpcalypso.wordpress.com/devdocs/design we can showcase different states of a component. For instance, here is the `JetpackProductCard` component presented with different props: https://wpcalypso.wordpress.com/devdocs/design/jetpack-product-card. This file can help you add a new configuration for that component, if you think it would be of interest to add a new one, `client/components/jetpack/card/jetpack-product-card/docs/example.tsx`.

elliottprogrammer replied:
Yes, good point, wouldn't want to forget to delete it. Ahh, cool!  It's like a UI component design system for all the calypso components!  Awesome! :)

elliottprogrammer commented:
> @elliottprogrammer, since you started this PR things have changed quite a bit in a couple of these files. Can you update your PR and ping us when it's ready for a new review?
> 
> Advice: always include Jetpack Voyager as reviewer so everyone can be aware of it.

@rcanepa , yes quite a lot has changed, thanks for the advice! Now I know to include Jetpack Voyager as reviewer. 👍

monsieur-z commented on client/components/jetpack/card/jetpack-product-card/index.tsx:
```diff
@@ -49,6 +49,7 @@ type OwnProps = {
 	isOwned?: boolean;
 	isDeprecated?: boolean;
 	expiryDate?: Moment;
+	showOptions?: boolean;
```
This one's not necessary anymore :)

monsieur-z commented on client/my-sites/plans-v2/plans-column/index.tsx:
```diff
@@ -102,6 +102,25 @@ const PlansColumn = ( { duration, onPlanClick, productType, siteId }: PlanColumn
+						2 && product.productSlug !== currentPlan ? (
+						<>
+							{ product.description }{ ' ' }
+							{ translate( '{{em}}Available options: Real-Time or Daily.{{/em}}', {
```
In the final copydeck, it's `Real-time`. This part could perhaps be extracted in a util function since it's used in both the product and plan column.

monsieur-z replied:
I originally meant just extracting what's inside the `<>`, since that's what's duplicated, but that's even better! Have two different functions is perfectly fine to me.

I feel that `utils` is a more adequate place than `translations`. I've stumbled upon the same problem lately: in that case, instead of renaming `utlils.ts` to `utils.tsx` (which could cause some annoying conflicts for now indeed), you could use the React functions directly instead of the JSX syntax, e.g. `React.createElement(React.Fragment, {}, children)`.

elliottprogrammer replied:
> you could use the React functions directly instead of the JSX syntax, e.g. React.createElement(React.Fragment, {}, children).

Oh yes, thats a great idea!  Will do, thanks!

monsieur-z commented on client/my-sites/plans-v2/plans-column/index.tsx:
```diff
@@ -102,6 +102,25 @@ const PlansColumn = ( { duration, onPlanClick, productType, siteId }: PlanColumn
+			.map( ( product ) => ( {
+				...product,
+				description:
+					product.subtypes?.filter( ( subtype ) => /_(daily|realtime)$/.test( subtype ) ).length >=
```
This regex needs to be updated because it would not work with monthly plans (e.g. `jetpack_security_daily_monthly`). Alternatively, you could try to get the plan object by its slug (`subtype` here) and check if an attribute indicates whether it's daily or realtime, but I don't know if such an attribute exists.

rcanepa replied:
Just an idea. What if we add a utility function that could answer the question by checking the `subtype` against a two lists of product slugs? I imagine one list could contain Daily products and the other Real-time products.

monsieur-z replied:
Seems more scalable to me indeed.

elliottprogrammer replied:
Ok, yes sounds good!  I've refactored it into a single function that compares against two lists of Daily and Real-time products. It should be much better now. :)

** Also I've since the PR got so far behind master (and there were a bunch of conflicts to resolve before it could be tested), [I moved the changes over to this new updated PR](https://github.com/Automattic/wp-calypso/pull/45197)

monsieur-z commented on client/my-sites/plans-v2/plans-column/index.tsx:
```diff
@@ -102,6 +102,25 @@ const PlansColumn = ( { duration, onPlanClick, productType, siteId }: PlanColumn
+				...product,
+				description:
+					product.subtypes?.filter( ( subtype ) => /_(daily|realtime)$/.test( subtype ) ).length >=
+						2 && product.productSlug !== currentPlan ? (
```
In the case of Jetpack Security, you could purchase Security _Daily_ or Security _Real-time_, but never Jetpack Security. I believe the condition should be updated to reflect that. Something like `! product.subtypes.includes(currentPlan)`.

monsieur-z commented:
It tested well for the plans, but I could still see the available options for Jetpack Backup with a site that has Backup Daily. Can you reproduce the issue?

The implementation is good. You might wanna check Renzo's comment which would make it more scalable I believe.

monsieur-z commented:
Just another detail. For products, we shouldn't display the available options if a user owns an option of another product in the same bundle. Right now, it doesn't change anything since the only bundle is Security, and the only product that has options is Backup, but we should probably leave a comment in the function.

elliottprogrammer commented:
> It tested well for the plans, but I could still see the available options for Jetpack Backup with a site that has Backup Daily. Can you reproduce the issue?
> 
> The implementation is good. You might wanna check Renzo's comment which would make it more scalable I believe.

Ok since the PR got so far behind master (and there were a bunch of conflicts to resolve before it could be tested), [I moved the changes over to this new updated PR](https://github.com/Automattic/wp-calypso/pull/45197)

I've refactored it into a single function that compares against two lists of Daily and Real-time products. It should be much better now, and more scalable. :)  And now with with the recent changes, it should be working correctly now (if i'm understanding it correctly).

> Just another detail. For products, we shouldn't display the available options if a user owns an option of another product in the same bundle. Right now, it doesn't change anything since the only bundle is Security, and the only product that has options is Backup, but we should probably leave a comment in the function.

Yes, this should be working as expected now..  (I think.., although I'm still having some trouble understanding this maybe). 
 The way it's working is, if you purchase Jetpack Security, all of the products that are bundled with Jetpack Security will not show available options, which the only product that offers available options right now is Backup, but even if any of the other products Did offer options (like Scan or something), then Scan would not show available options either, because it's already owned/bundled under Jetpack security. Hmm, but maybe I'm still not quite getting it?  Let me know. 👍  :)

monsieur-z commented:
> Let me know if I'm still not quite understanding correctly.

Let's maybe ping Yvonne in the new PR for review, @elliottprogrammer. It's a bit different from what I understood initially but there's been some changes since the start of the project so let's take a look at it with a fresh pair of eyes.

monsieur-z commented:
Can we close this PR, @elliottprogrammer?

elliottprogrammer commented:
> Can we close this PR, @elliottprogrammer?

Yes!  Done! 👍
